### PR TITLE
REGR: Fixed truncation with na_rep

### LINKED ISSUE #31447: REGR: to_csv(na_rep) truncates values
#### Code Sample

```python
>>> import pandas as pd
>>> pd.__version__
'1.0.0rc0'
>>> pd.Series(range(8, 12)).to_csv(na_rep='-')
',0\r\n0,8\r\n1,9\r\n2,1\r\n3,1\r\n'
```
#### Problem description

Despite (or due to?) #29975 and #30146, using `na_rep` still makes `to_csv()` truncate values. In this case integers.

#### Expected Output

',0\r\n0,8\r\n1,9\r\n2,10\r\n3,11\r\n'

#### Output of ``pd.show_versions()``

<details>
INSTALLED VERSIONS
------------------
commit           : None
python           : 3.8.1.final.0
python-bits      : 64
OS               : Windows
OS-release       : 10
machine          : AMD64
processor        : Intel64 Family 6 Model 142 Stepping 12, GenuineIntel
byteorder        : little
LC_ALL           : None
LANG             : None
LOCALE           : Dutch_Netherlands.1252

pandas           : 1.0.0rc0
numpy            : 1.18.1
pytz             : 2019.3
dateutil         : 2.8.1
pip              : 20.0.2
setuptools       : 41.2.0
Cython           : None
pytest           : None
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 2.10.3
IPython          : 7.11.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
gcsfs            : None
lxml.etree       : None
matplotlib       : 3.1.2
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pytables         : None
pytest           : None
s3fs             : None
scipy            : 1.4.1
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
xlwt             : None
xlsxwriter       : None
numba            : None
</details>

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/31447

### CODE REVIEW

File: pandas/core/internals/blocks.py [hunk 1]
```diff
@@ -684,7 +684,10 @@ def to_native_types(self, slicer=None, na_rep="nan", quoting=None, **kwargs):
         itemsize = writers.word_len(na_rep)
 
         if not self.is_object and not quoting and itemsize:
-            values = values.astype(f"<U{itemsize}")
+            values = values.astype(str)
+            if values.dtype.itemsize / np.dtype("U1").itemsize < itemsize:
```
Comment (WillAyd) [inline pandas/core/internals/blocks.py]: Any performance implications to this? An alternate strategy I think would be to compare to the `itemsize` to the length of the `na_rep` and take the max, which I think would reduce the amount of `astype` calls needed, assuming they come at a noticeable cost

Comment (TomAugspurger) [reply pandas/core/internals/blocks.py]: Can you reword your suggestion? I don't quite follow. In particular, I think the variable `itemsize` here *is* the length of the stringified na_rep.

My idea here was: we don't know what the string lengths will be without scanning the values and doing the actually conversion. So just let NumPy do that, and only astype to a larger dtype if needed.

Comment (WillAyd) [reply pandas/core/internals/blocks.py]: Hmm sorry I was mistakenly thinking that the issue was only truncating when all values in the array had a str length that was less than that of na_rep, which isn't the case

Do we even need the second astype?

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @TomAugspurger

**Pull request merged**